Add blobtk depth for read coverage

sainsachiko opened the pull request:
- [x] Add blobtk depth for read coverage

<!--
# sanger-tol/readmapping pull request

Many thanks for contributing to sanger-tol/readmapping!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/sanger-tol/readmapping/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/sanger-tol/readmapping/tree/master/.github/CONTRIBUTING.md)
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

sainsachiko committed:
Add blobtk depth to calculate read coverage

sanger-tolsoft commented:
<!-- add-pr-comment:add-pr-comment -->

> [!WARNING]
> Newer version of the nf-core template is available.
>
> Your pipeline is using an old version of the nf-core template: 3.2.1.
> Please update your pipeline to the latest version.
>
> For more documentation on how to update your pipeline, please see the [nf-core documentation](https://github.com/nf-core/tools?tab=readme-ov-file#sync-a-pipeline-with-the-template) and [Synchronisation documentation](https://nf-co.re/docs/contributing/sync).

muffato commented on conf/base.config:
```diff
@@ -133,4 +133,9 @@ process {
     }
+
+    withName: BLOBTK_DEPTH {
+        cpus   = { log_increase_cpus(4, 2*task.attempt, 1, 2) }
```
You can remove this line because the tool is single-threaded and the default `cpus` is already 1
```suggestion
```

muffato commented on modules/local/blobtk_depth.nf:
```diff
@@ -0,0 +1,34 @@
+    blobtk depth \\
+        -c ${cram} \\
+        $args \\
+        -O ${prefix}.regions.bed.gz \\
```
As far as I can see this is a regular gzip file.
To allow range queries, it should rather be bgzip compression. Can you add https://github.com/nf-core/modules/blob/master/modules/nf-core/samtools/bgzip/main.nf after it to convert from regular gzip to bgzip ?

sainsachiko replied:
Hi @muffato, I have added Bgzip in [8e2a7cc](https://github.com/sanger-tol/readmapping/pull/171/commits/8e2a7ccd771c09a40f23c2c16a59014eb509db00). Please kindly review. Thank you!

muffato commented on modules/local/blobtk_depth.nf:
```diff
@@ -0,0 +1,34 @@
+    blobtk depth \\
+        -c ${cram} \\
+        $args \\
+        -O ${prefix}.regions.bed.gz \\
```
The format is actually bedGraph, not bed. And I think "coverage" would be clearer than "regions"
```suggestion
        -O ${prefix}.coverage.bedGraph.gz \\
```

muffato commented on docs/output.md:
```diff
@@ -89,6 +89,19 @@ The filtered PacBio reads are aligned with `MINIMAP2_ALIGN`. The sorted and merg
 
+### Read coverage
+
+Read coverage of the output alignment file is calculated with [blotk_depth](https://github.com/genomehubs/blobtk/wiki/blobtk-depth).
```
```suggestion
Read coverage of the output alignment file is calculated with [blobtk depth](https://github.com/genomehubs/blobtk/wiki/blobtk-depth).
```

muffato commented on docs/output.md:
```diff
@@ -89,6 +89,19 @@ The filtered PacBio reads are aligned with `MINIMAP2_ALIGN`. The sorted and merg
+
+- `read_mapping`
+  - `<sequence-type>`
+    - `<gca_accession>.unmasked.<sequence-type>.<sample_id>.[cr|b]am.regions.bed.gz`: Read coverage in BED format
```
```suggestion
    - `<gca_accession>.unmasked.<sequence-type>.<sample_id>.[cr|b]am.coverage.bedGraph.gz`: Read coverage in bedGraph format
```

sainsachiko committed:
Apply suggestions from code review

- Remove cpus limit as blobtk depth is single-threaded
- Adjust coverage file name
- Fix typos

Co-authored-by: Matthieu Muffato <mm49@sanger.ac.uk>

sainsachiko committed:
Add BGZIP to compress bedGraph coverage

sainsachiko committed:
Update tabix/bgzip

muffato commented on conf/base.config:
```diff
@@ -133,4 +133,9 @@ process {
     }
+
+    withName: 'BLOBTK_DEPTH|BGZIP_BEDGRAPH' {
+        memory = { 1.GB  * Math.ceil( meta.read_count / 1000000 ) * task.attempt }
```
I don't know how much memory they use exactly, but I think BLOBTK_DEPTH depends more on the genome size than the number of reads, and BGZIP is just compressing the file by blocks so its memory usage should be constant.
In any case, I think we have sequencing runs with more than a billion reads, so let's just remove the dependency on `meta.read_count`. We'll have a first taste of real world usage through the full test.

```suggestion
        memory = { 4.GB * task.attempt }
```

muffato reviewed (commented):
Can you update the CHANGELOG too ?

I can reproduce the issue with `-resume` you were saying this morning. Did you find out why ? If you run Nextflow with `-dump-hashes` you can find out what changes between runs, cf <https://www.nextflow.io/docs/latest/cache-and-resume.html#comparing-the-hashes-of-two-runs>

sainsachiko committed:
Fix: BAM/CRAM reheader and BLOBTK DEPTH cache

sainsachiko committed:
Use ln -s to change file names

sainsachiko committed:
Merge branch 'dev' into hhoang-read-coverage

sainsachiko commented:
@sanger-tolsoft fix linting

sainsachiko committed:
Fix linting

sainsachiko committed:
Remove module config of SAMTOOLS_REINDEX

muffato committed:
Not used

muffato committed:
We pass a BAM file, not a CRAM file

muffato committed:
Too many names like "ch_bams" is confusing !

muffato committed:
Simplified the CHANGE_NAME module

- `process_*` don't exist in this pipeline. The defaults resources are already for single CPU use
- No need to invoke a container, the host must have `ln`
- Therefore no need to report versions

muffato committed:
It's the CRAM output that we want

muffato committed:
No conversion is happening here

muffato committed:
Allow rerunning

muffato committed:
Our convention is not to prefix workflow inputs with "ch_"

muffato committed:
More straightforward sub-workflow

- CHANGE_NAME is always called and is the only place where the published
  file name is set
- All downstream modules use the renamed BAMs as input and copy its name
  to the output

muffato committed:
Trick for not publishing the BAM when CRAM output is selected

We don't publish anything from CHANGE_NAME itself. Instead we make
SAMTOOLS_INDEX publish its input file.

muffato committed:
More explicit name

muffato committed:
Merge pull request #176 from sanger-tol/hhoang-read-coverage_review

Read coverage review

sainsachiko committed:
Input both file and index for stats process

sainsachiko committed:
Publish cram index when provide header

muffato committed:
Remove trailing whitespace

muffato committed:
Merge pull request #177 from sanger-tol/hhoang-read-coverage_change_name

"change_name" process and configuration of output file name

muffato committed:
Remove a trailing empty line

muffato reviewed (changes requested):
I've just found there's an issue with the BAM output !
When running this branch with the `--outfmt bam` option, the output (BAM) files only contain the header, not any read alignment.

muffato commented:
And another issue is that `pacbio` samples (i.e. `mMelMel3_T3` and `mMelMel3_T4`) go through `CONVERT_STATS` **twice**. They may be going twice earlier than that but I'm not sure. I may be confused by the chunking.
```
$ grep CONVERT_STATS:SAMTOOLS_FLAGSTAT results/pipeline_info/execution_trace_2025-07-15_10-54-16.txt | cut -f4 | awk '{print $2}' | sort | uniq -c
      1 (mMelMel1_T1)
      1 (mMelMel2_T1)
      1 (mMelMel3_T1)
      1 (mMelMel3_T2)
      2 (mMelMel3_T3)
      2 (mMelMel3_T4)
```

sainsachiko committed:
Merge pacbio chunked bam after alignment

sainsachiko committed:
Do not publish BAM from SAMTOOLS_INDEX if re-header BAMs

muffato committed:
[lint] Removed trailing whitespace

muffato commented:
I've run some tests (CRAM/BAT/both, and with/without header). No issue about empty BAM file like this morning.

In BAM + no-header mode, the BAM files are not published (but the BAI are)
```
results_coverage8_noh_bam/read_mapping/hic
results_coverage8_noh_bam/read_mapping/hic/GCA_922984935.2.subset.unmasked.hic.mMelMel3_T1.flagstat
results_coverage8_noh_bam/read_mapping/hic/GCA_922984935.2.subset.unmasked.hic.mMelMel3_T1.coverage.bedGraph.gz
results_coverage8_noh_bam/read_mapping/hic/GCA_922984935.2.subset.unmasked.hic.mMelMel3_T1.stats
results_coverage8_noh_bam/read_mapping/hic/GCA_922984935.2.subset.unmasked.hic.mMelMel3_T1.idxstats
results_coverage8_noh_bam/read_mapping/hic/GCA_922984935.2.subset.unmasked.hic.mMelMel3_T1.bam.bai
```

(to enable the no-header mode, I edit `conf/test.config` and comment out the line)

sainsachiko committed:
Publish bam file when no header

muffato committed:
[lint] Removed trailing whitespace
